rate_limit.py

```diff
@@ -0,0 +1,97 @@
+"""
+This module contains functions for rate limiting requests.
+"""
+
+import datetime
+import signal
+import sys
+from uuid import uuid4
+
+from apscheduler.schedulers import background
+import gradio as gr
+
+
+class InvalidTokenException(Exception):
+  pass
+
+
+class UserRateLimitException(Exception):
+  pass
+
+
+class SystemRateLimitException(Exception):
+  pass
+
+
+class RateLimiter:
+
+  def __init__(self, daily_limit=10000):
+    self.requests = {}
+    self.request_count = 0
+    self.daily_limit = daily_limit
+    self.scheduler = background.BackgroundScheduler()
+
+  def request_allowed(self, token: str):
+    if not token or token not in self.requests:
+      raise InvalidTokenException()
+
+    if (datetime.datetime.now() - self.requests[token]).seconds < 5:
+      raise UserRateLimitException()
+
+    if self.request_count >= self.daily_limit:
+      raise SystemRateLimitException()
+
+    self.requests[token] = datetime.datetime.now()
+    self.request_count += 1
+
+  def initialize_request(self, token: str):
+    self.requests[token] = datetime.datetime.min
+
+  def clean_up(self):
+    for token, last_request_time in dict(self.requests).items():
+      if (datetime.datetime.now() - last_request_time).days >= 1:
+        del self.requests[token]
+
+  def reset_request_count(self):
+    self.request_count = 0
+
+
+rate_limiter = RateLimiter()
+
+
+def set_token(app: gr.Blocks):
+
+  def set_token_server():
```

Comment:
https://google.github.io/styleguide/pyguide.html#26-nestedlocalinner-classes-and-functions
just fyi, google style guide suggests to not nest functions unless they close local variables.

Reply:
thank you for the info. maybe we need an exception here because Gradio components receive functions as arguments so it’s more like.. defining a variable I think 🤔